spark-cassandra-connector/src/main/scala/com/datastax/spark/connector/writer/Batch.scala

```diff
@@ -0,0 +1,76 @@
+package com.datastax.spark.connector.writer
+
+import com.datastax.driver.core.BoundStatement
+import com.datastax.spark.connector.{BatchSize, BytesInBatch, RowsInBatch}
+
+import scala.collection.mutable.ArrayBuffer
+
+/** A simple wrapper over a collection of bound statements. */
+private[writer] sealed trait Batch extends Ordered[Batch] {
+  def add(stmt: BoundStatement): Unit
+
+  /** Returns `true` if the collected statements exceed the desired size of this batch. Rows based and
+    * bytes based implementations compute it in different ways. */
+  def isSizeExceeded: Boolean
+
+  def statements: Seq[BoundStatement]
+
+  /** Removes all the collected statements and resets this batch to the initial state. */
+  def reuse(): Unit
+}
+
+private[writer] object Batch {
+  implicit val batchOrdering = new Ordering[Batch] {
+    override def compare(x: Batch, y: Batch): Int = x.compare(y)
+  }
+
+  def apply(batchSize: BatchSize): Batch = {
+    batchSize match {
+      case RowsInBatch(rows) => new RowLimitedBatch(rows)
+      case BytesInBatch(bytes) => new SizeLimitedBatch(bytes)
+    }
+  }
+}
+
+private[writer] class RowLimitedBatch(val maxRows: Int) extends Batch {
+  val buf = new ArrayBuffer[BoundStatement](maxRows)
+
+  def add(stmt: BoundStatement): Unit = {
+    buf += stmt
```

Comment:
This limit is not intended to be strict. This is only a hint. However, I found that I should allocate `maxRows+1` for `buf`.